### PR TITLE
Alerting: Use time_intervals instead of the deprecated mute_time_intervals in a…

### PR DESCRIPTION
**Pull request by soniaAguilarPeiron**
**What is this feature?**

This PR updates the way we use time intervals , after the [the deprecation of mute_time_intervals](https://prometheus.io/docs/alerting/latest/configuration/#time_interval) from the alert manager config.

In this PR:

**Grafana Alertmanager**

Use  `mute_time_intervals` and `time_intervals` when showing the UI, and save only on `mute_time_intervals`

**External Alertmanager**

- Combine both `mute_time_intervals` and `time_intervals` when showing the UI.
- Use `time_intervals` when creating new time intervals in the UI.
- When editing an existing time interval, if it came from a `mute_time_interval` save it back as a `mute_time_interval`, and if it came from a `time_interval `save it back as a `time_interval `

Fixes part of this [issue](https://github.com/grafana/grafana/issues/83168)

**Why do we need this feature?**

https://github.com/grafana/grafana/issues/83168

**Who is this feature for?**

All users.

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Comment by grobinson-grafana** [comment]
I think I found a bug. If there is mixed use of `time_intervals` and `mute_time_intervals` (for example, because `time_intervals` were created in the UI, but there are still provisioned `mute_time_intervals`), then the `time_intervals` are not shown until all `mute_time_intervals` are deleted:

Screenshot 2024-02-22 at 4 39 43 PM
Screenshot 2024-02-22 at 4 40 02 PM

**Comment by grobinson-grafana** [comment]
I tested it and LGTM! I'll let @gillesdemey and @konrad147 do the code review and approval.

File: public/app/features/alerting/unified/NotificationPolicies.tsx [hunk 1]
```diff
@@ -191,8 +191,11 @@ const AmRoutes = () => {
   if (!selectedAlertmanager) {
     return null;
   }
-
-  const numberOfMuteTimings = result?.alertmanager_config.mute_time_intervals?.length ?? 0;
+  const time_intervals = [
+    ...(result?.alertmanager_config?.mute_time_intervals ?? []),
+    ...(result?.alertmanager_config?.time_intervals ?? []),
+  ];
+  const numberOfMuteTimings = time_intervals?.length ?? 0;
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/NotificationPolicies.tsx:198]: nit: I don't think `time_intervals` can be undefined when we spread in to a new Array

File: public/app/features/alerting/unified/components/mute-timings/MuteTimingForm.tsx [hunk 2]
```diff
@@ -70,19 +78,51 @@ const MuteTimingForm = ({ muteTiming, showError, loading, provenance }: Props) =
 
     const newMuteTiming = createMuteTiming(values);
 
-    const muteTimings = muteTiming
-      ? config?.mute_time_intervals?.filter(({ name }) => name !== muteTiming.name)
-      : config?.mute_time_intervals;
-
+    const originalMuteTimingsWithoutNew = fromMuteTimings
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/components/mute-timings/MuteTimingForm.tsx:81]: Can we write a separate function like `replaceMuteTiming` to handle the omit + re-add logic?

File: public/app/features/alerting/unified/state/actions.ts [hunk 3]
```diff
@@ -693,10 +693,24 @@ export const deleteMuteTimingAction = (alertManagerSourceName: string, muteTimin
       alertmanagerApi.endpoints.getAlertmanagerConfiguration.initiate(alertManagerSourceName)
     ).unwrap();
 
-    const muteIntervals =
-      config?.alertmanager_config?.mute_time_intervals?.filter(({ name }) => name !== muteTimingName) ?? [];
+    const isGrafanaDatasource = alertManagerSourceName === GRAFANA_RULES_SOURCE_NAME;
+
+    const muteIntervalsFiltered =
+      (config?.alertmanager_config?.mute_time_intervals ?? [])?.filter(({ name }) => name !== muteTimingName) ?? [];
+    const time_intervalsFiltered =
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/state/actions.ts:700]: nit: variable naming casing is inconsistent here :)

File: public/app/features/alerting/unified/NotificationPolicies.tsx (hunk 1 above)
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/NotificationPolicies.tsx:198]: 💯

File: public/app/features/alerting/unified/hooks/useMuteTimingOptions.ts [hunk 4]
```diff
@@ -13,8 +13,10 @@ export function useMuteTimingOptions(): Array<SelectableValue<string>> {
   const config = currentData?.alertmanager_config;
 
   return useMemo(() => {
+    // merge both fields mute_time_intervals and time_intervals to support both old and new config
+    const time_intervals = [...(config?.mute_time_intervals ?? []), ...(config?.time_intervals ?? [])];
```
Comment (konrad147) [inline public/app/features/alerting/unified/hooks/useMuteTimingOptions.ts:17]: Should we create a function which merges time interval from the config? The same merging code is repeated in different places

File: public/app/features/alerting/unified/components/mute-timings/MuteTimingForm.tsx [hunk 5]
```diff
@@ -12,14 +12,16 @@ import { useAlertmanager } from '../../state/AlertmanagerContext';
...
+import { GRAFANA_RULES_SOURCE_NAME } from '../../utils/datasource';
 import { makeAMLink } from '../../utils/misc';
 import { createMuteTiming, defaultTimeInterval } from '../../utils/mute-timings';
 import { ProvisionedResource, ProvisioningAlert } from '../Provisioning';
 
 import { MuteTimingTimeInterval } from './MuteTimingTimeInterval';
 
 interface Props {
-  muteTiming?: MuteTimeInterval;
+  fromMuteTimings?: MuteTimeInterval; // mute time interval when comes from the old config , mute_time_intervals
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/mute-timings/MuteTimingForm.tsx:23]: Good idea. I prefer keeping the from prefix though `fromLegacyTimeInterval`

File: public/app/features/alerting/unified/components/mute-timings/MuteTimingsTable.tsx [hunk 6]
```diff
@@ -72,9 +72,10 @@ export const MuteTimingsTable = ({ alertManagerSourceName, muteTimingNames, hide
   const config = currentData?.alertmanager_config;
 
   const [muteTimingName, setMuteTimingName] = useState<string>('');
-
   const items = useMemo((): Array<DynamicTableItemProps<MuteTimeInterval>> => {
-    const muteTimings = config?.mute_time_intervals ?? [];
+    // merge both fields mute_time_intervals and time_intervals to support both old and new config
+    const time_intervals = [...(config?.mute_time_intervals ?? []), ...(config?.time_intervals ?? [])];
+    const muteTimings = time_intervals ?? [];
```
[the remark this answers is not in the document]
Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/components/mute-timings/MuteTimingsTable.tsx:78]: good catch

File: public/app/features/alerting/unified/MuteTimings.tsx [hunk 7]
```diff
@@ -21,8 +21,10 @@ const MuteTimings = () => {
   const config = currentData?.alertmanager_config;
 
   const getMuteTimingByName = useCallback(
-    (id: string): MuteTimeInterval | undefined => {
-      const timing = config?.mute_time_intervals?.find(({ name }: MuteTimeInterval) => name === id);
+    (id: string, fromTimeIntervals: boolean): MuteTimeInterval | undefined => {
+      // merge both fields mute_time_intervals and time_intervals to support both old and new config
+      const time_intervals = fromTimeIntervals ? config?.time_intervals ?? [] : config?.mute_time_intervals ?? [];
+      const timing = time_intervals?.find(({ name }: MuteTimeInterval) => name === id);
```
Comment (gillesdemey) [inline public/app/features/alerting/unified/MuteTimings.tsx:27]: nit: optional chaining not needed here, time_intervals should be an array always :)

Comment (soniaAguilarPeiron) [reply public/app/features/alerting/unified/MuteTimings.tsx:27]: ohhhh yes, I forgot this one

**Review by gillesdemey — APPROVED** [review]

**Pull request merged**